declarativewidgets/proposal.md

```diff
@@ -0,0 +1,68 @@
+
+### Scope
+The objective of this incubation project is to establish a baseline prototype that begins to blur the lines between a notebook and an application while bootstrapping the ecosystem of notebook compatible widgets. It will build on top of the foundation set by IPywidgets.
+
+This incubation effort can be scoped as:
+
+1. Identification of the minimum essential pieces of IPyWidgets that need to exist in all language kernels
+2. Building a portable layer can leverage an existing and ever growing ecosystem of web widgets.
+3. Providing a widget framework that works across the various and growing set of languages that can be hosted in a Jupyter Interactive Notebook. 
+
+### Functionality
+
+As outlined in this [blog post](http://blog.ibmjstart.net/2015/08/21/declarative-widget-system-for-jupyter-notebooks/), the intent of this proposal is to provide support for:
+
+* A **function element** which will allow the running of a segment of code within the notebook as a side effect of a user's interactivity with UI elements. By using a function defined in a code cell, the author can connect other visual elements to set the function arguments, invoke it and display its result.
+* A Pandas or a Spark DataFrame defined is a way to interact with data. Using a **dataframe element** the author can bind other visual elements to display the data held by a dataFrame. The dataframe element can be set to listen to changes to the DataFrame instance in the kernel and trigger notifications that change other elements that are bound to it.
+* Extensions to the data binding capabilities of Polymer to allow multiple cell output areas to share data and related events. Support for changing and watching data using the language of the kernel.
+* An element and related server extension to support importing new 3rd party elements into the Notebook. The import mechanism used bower to install new Polymer element upon user's request.
+ 
+## Audience
+
+1. Developers of rich interactive web widgets.
+2. Notebook users who desire to inject rich interactive widgets into their notebooks.
+3. Notebook users who wish to create interactive dashboards in their notebooks.
+
+## Other options
+The [IPyWidgets](https://github.com/ipython/ipywidgets) project is the reference implementation of interactive widgets for the Python language. At its core, it defines a communication channel and protocol between the browser and the Python kernel. This channel is used to exchange messages between a Javascript and a Python representation of a widget. It keeps the state of the two in sync and triggers events. A user clicks on a button in the browser, a message is sent to the Python representation of the button and it performs a preconfigured action. However, the IPyWidgets project is not architected to achieve the compatibility and portability goals associated Jupyter language kernels.
```

Comment:
I view ipywidgets as both an *other option* but also one of the main integration targets. We have lots of design questions and work around ipywidgets right now that are ongoing and which would overlap with some of the ideas here. I don't think this is bad - we really want ipywidgets to evolve and be useful across all languages and frontend frameworks (polymer, react, etc.). It will be fun to think about how we can pull together the best of all these ideas.